clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -698,36 +753,50 @@ synchronized boolean hasUnresolvedSequences() {
     }
 
     synchronized boolean hasUnresolvedSequence(TopicPartition topicPartition) {
-        return partitionsWithUnresolvedSequences.contains(topicPartition);
+        return partitionsWithUnresolvedSequences.containsKey(topicPartition);
     }
 
-    synchronized void markSequenceUnresolved(TopicPartition topicPartition) {
-        log.debug("Marking partition {} unresolved", topicPartition);
-        partitionsWithUnresolvedSequences.add(topicPartition);
+    synchronized void markSequenceUnresolved(ProducerBatch batch) {
+        log.debug("Marking partition {} unresolved", batch.topicPartition);
+        partitionsWithUnresolvedSequences.put(batch.topicPartition, batch.baseSequence() + batch.recordCount);
     }
 
     // Checks if there are any partitions with unresolved partitions which may now be resolved. Returns true if
     // the producer id needs a reset, false otherwise.
-    private boolean shouldResetProducerStateAfterResolvingSequences() {
-        for (Iterator<TopicPartition> iter = partitionsWithUnresolvedSequences.iterator(); iter.hasNext(); ) {
+    public void checkUnresolvedSequences() {
+        for (Iterator<TopicPartition> iter = partitionsWithUnresolvedSequences.keySet().iterator(); iter.hasNext(); ) {
             TopicPartition topicPartition = iter.next();
             if (!hasInflightBatches(topicPartition)) {
```

Comment:
You're right about the problem with the logic. The transition to fatal now happens in this method, so the check in `Sender.runOnce` is unnecessary. Either I had neglected to remove it before, or I pulled it back in when I merged your refactor.

Regarding aborting in-flight batches, I think we want to wait for the inflights because it might wind up resolving all the sequences, in which case we wouldn't actually want to abort. If we tried to fast-abort here, I think we would wind up aborting (potentially unnecessarily) every time a response came back out of order.

Reply:
(Deleted an earlier comment that was inaccurate)

Yes, you are correct about the sender logic. That transition to fatal shouldn't be there, I either neglected to remove it or added it back when I merged in your refactor. I removed it and added a test to catch that case.

You are also right about the in-flight batches, we could abort them early. I'd rather do that in a separate change, though.